File: test/java/src/test/java/id/dana/util/DisbursementCustomerRetry.java
package id.dana.util;

import id.dana.invoker.model.exception.DanaException;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public final class DisbursementCustomerRetry {

  public static final List<String> CUSTOMER_NUMBERS = Collections.unmodifiableList(
      Arrays.asList(
          "62811742234",
          "62817345544",
          "62817345545"
      )
  );

  private static final Pattern RESPONSE_CODE_PATTERN =
      Pattern.compile("\"responseCode\"\\s*:\\s*\"((?:403|404)[^\"]*)\"");

  private DisbursementCustomerRetry() {}

  public static boolean isForbiddenResponseCode(String code) {
    return code != null && (code.startsWith("403") || code.startsWith("404"));
  }

  public static boolean isForbiddenException(Throwable exception) {
    String message = exception.getMessage();
    if (message == null) {
      return false;
    }
    Matcher matcher = RESPONSE_CODE_PATTERN.matcher(message);
    if (matcher.find()) {
      return isForbiddenResponseCode(matcher.group(1));
    }
    return message.contains("403") || message.contains("404");
  }

  @FunctionalInterface
  public interface CustomerNumberOperation<T> {
    T apply(String customerNumber) throws Exception;
  }

  @FunctionalInterface
  public interface ResponseCodeExtractor<T> {
    String extract(T result);
  }

  public static <T> RetryResult<T> withCustomerNumberRetry(
      CustomerNumberOperation<T> operation,
      ResponseCodeExtractor<T> responseCodeExtractor
  ) throws Exception {
    Exception lastException = null;
    for (String customerNumber : CUSTOMER_NUMBERS) {
      try {
        T result = operation.apply(customerNumber);
        String responseCode = responseCodeExtractor.extract(result);
        if (isForbiddenResponseCode(responseCode)) {
          lastException = new DanaException("responseCode=" + responseCode);
          continue;
        }
        return new RetryResult<>(result, customerNumber);
      } catch (Exception exception) {
        if (isForbiddenException(exception)) {
          lastException = exception;
          continue;
        }
        throw exception;
      }
    }
    if (lastException != null) {
      throw lastException;
    }
    throw new IllegalStateException("All customer numbers returned 403/404");
  }

  public static final class RetryResult<T> {
    private final T result;
    private final String customerNumber;

    public RetryResult(T result, String customerNumber) {
      this.result = result;
      this.customerNumber = customerNumber;
    }

    public T result() {
      return result;
    }

    public String customerNumber() {
      return customerNumber;
    }
  }
}

File: test/php/helper/DisbursementCustomerRetry.php
<?php

namespace DanaUat\Helper;

use Dana\ApiException;

class DisbursementCustomerRetry
{
    public const CUSTOMER_NUMBERS = [
        '62811742234',
        '62817345544',
        '62817345545',
    ];

    public static function isForbiddenResponseCode(?string $code): bool
    {
        return $code !== null && (str_starts_with($code, '403') || str_starts_with($code, '404'));
    }

    public static function isForbiddenException(\Exception $e): bool
    {
        if (in_array((int) $e->getCode(), [403, 404], true)) {
            return true;
        }
        if ($e instanceof ApiException) {
            $body = $e->getResponseBody();
            if (is_string($body) && preg_match('/"responseCode"\s*:\s*"((?:403|404)[^"]*)"/', $body, $matches)) {
                return self::isForbiddenResponseCode($matches[1]);
            }
            if (is_object($body) && isset($body->responseCode)) {
                return self::isForbiddenResponseCode((string) $body->responseCode);
            }
        }
        $message = $e->getMessage();
        if (preg_match('/"responseCode"\s*:\s*"((?:403|404)[^"]*)"/', $message, $matches)) {
            return self::isForbiddenResponseCode($matches[1]);
        }
        return str_contains($message, '403') || str_contains($message, '404');
    }

    /**
     * @template T
     * @param callable(string): T $operation
     * @param callable(T): ?string|null $getResponseCode
     * @return array{0: T, 1: string}
     */
    public static function withCustomerNumberRetry(callable $operation, ?callable $getResponseCode = null): array
    {
        $lastException = null;
        foreach (self::CUSTOMER_NUMBERS as $customerNumber) {
            try {
                $result = $operation($customerNumber);
                $code = $getResponseCode ? $getResponseCode($result) : null;
                if ($code === null && is_object($result) && method_exists($result, 'getResponseCode')) {
                    $code = $result->getResponseCode();
                }
                if (self::isForbiddenResponseCode($code)) {
                    $lastException = new \RuntimeException("responseCode={$code}");
                    continue;
                }
                return [$result, $customerNumber];
            } catch (\Exception $e) {
                if (self::isForbiddenException($e)) {
                    $lastException = $e;
                    continue;
                }
                throw $e;
            }
        }
        if ($lastException !== null) {
            throw $lastException;
        }
        throw new \RuntimeException('All customer numbers returned 403/404: ' . implode(', ', self::CUSTOMER_NUMBERS));
    }
}
